spk init generates more fields & docs in sandstorm-pkgdef.capnp

zarvox opened the pull request:
We now generate a stub manifest for developers to fill out, and a bridgeConfig
they can uncomment to specify permissions and roles more easily.

We also now provide numerous links to the .capnp files where more detailed
comments and guidance live.

Fixes #787.

ocdtrekkie commented:
\o/

kentonv commented:
Can you make sure there is a blank line between each comment and the next field declaration? This is particularly important to clarify our comment style where the comments appear after the declarations. :)

ocdtrekkie commented:
I know there's still tweaks to be made to my docs for the publishing guide, but shouldn't we refer people to docs rather than a source code file in the comments?

mquandalle commented on src/sandstorm/package.capnp:
```diff
@@ -365,7 +365,7 @@ struct Metadata {
     # that is not susceptible to spoofing or forgery. Therefore, we *only* identify the author by
     # PGP key. Various PGP infrastructure exists which can be used to determine the author's
-    # identity based on their PGP key. For exmaple, Keybase.io has done a really good job of
+    # identity based on their PGP key. For example, Keybase.io has done a really good job of
```
I fixed another instance of this type in #800, and greping it shows a remaining one in `./deps/capnproto/c++/src/kj/units.h:`

zarvox commented:
@kentonv @ocdtrekkie Addressed comments, thanks. :)

@paulproteus I just realized adding the manifest section as-is (with some `embed "file/that/does/not/exist"` stanzas) will probably break the packaging tutorial, since `spk dev` and `spk pack` will be unhappy about embedding a missing file.  Do you think it's better to have those fields that we usually embed commented out by default, or should we adjust the packaging tutorial to require those fields?  I lean toward the former, but wanted to get your opinion.

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "      # https://github.com/sandstorm-io/sandstorm/blob/master/src/sandstorm/package.capnp\n"
+        "      icons = (\n"
+        "        # Various icons to represent the app in various contexts.\n"
+        "        appGrid = (svg = embed \"path/to/128x128.svg\"),\n"
```
Suggest: Comment out the icons to avoid it not compiling; use `appGrid` in the appGrid*svg filename to indicate the purpose of the icon

paulproteus replied:
Request: generate identicons as svg here as part of spk init

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "      categories = [],\n"
+        "      # A list of categories/genres to which this app belongs, sorted with best fit first.\n"
+        "      # See the list of categories in\n"
+        "      # https://github.com/sandstorm-io/sandstorm/blob/master/src/sandstorm/package.capnp\n"
```
Suggest: Link to https://docs.sandstorm.io/en/latest/developing/publishing-apps/

paulproteus replied:
i,i write the ducks

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "        # requests as well as app store administrator requests, so it is very important that this be a\n"
+        "        # valid address with someone paying attention to it.\n"
+        "\n"
+        "        pgpSignature = embed \"path/to/pgp-signature\",\n"
```
Suggest: comment this out for now

paulproteus replied:
In the future it'd be cool to have some smarter integration

paulproteus replied:
Suggest:

```
        "        # You can create a signature file using `gpg` like so:\n"
        "        #\n"
        "        #     echo -n \"I am the author of the Sandstorm.io app with the following ID: <app-id>\" | gpg --sign > gp-signature\n"
        "        #\n"
        "        # Further details including how to set up GPG and how to use keybase.io are in the docs.sandstorm.io page about app publishing."
```

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "        #\n"
+        "        #     echo -n \"I am the author of the Sandstorm.io app with the following ID: <app-id>\" | gpg --sign > pgp-signature\n"
+        "        #\n"
+        "        # To learn how to set up gpg, visit Keybase (https://keybase.io) -- they have excellent\n"
```
Suggest moving the paragraphs where possible to be links to the docs, but I'm +1 on keeping essential copy-pasta in here as needed.

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "      # for purposes like marketing and display.  These fields are documented at\n"
+        "      # https://docs.sandstorm.io/en/latest/developing/publishing-apps/#add-required-metadata\n"
+        "      # and in the source code in the Metadata section of\n"
+        "      # https://github.com/sandstorm-io/sandstorm/blob/master/src/sandstorm/package.capnp\n"
```
-0 on source code link here; the docs should be enough and if they're not we should fix that IMHO (but if you think it's essential I can live)

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "        # the Sandstorm app install flow and app market will present this information to the user as\n"
+        "        # \"verified\".\n"
+        "\n"
+        "        #upstreamAuthor = \"Example App Team\",\n"
```
Why this one commented out but the packager not commented out?

I think uncomment this one!

paulproteus replied:
I guess uncomment this and say, "Remove this line if you consider yourself the author of the app."

Also remark that "If you omit this, you'll show up as the upstream author on the app market"

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "        # signature only \"packaged\" the app for Sandstorm, rather than developing the app.\n"
+        "      ),\n"
+        "      pgpKeyring = embed \"path/to/pgp-keyring\",\n"
+        "      # A keyring in GPG keyring format containing all public keys needed to verify PGP signatures in\n"
```
Remove this paragraph IMHO

paulproteus replied:
But if you want to keep it, I can live with it.

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "        # produced the Sandstorm package. Setting this implies that the author connected to the PGP\n"
+        "        # signature only \"packaged\" the app for Sandstorm, rather than developing the app.\n"
+        "      ),\n"
+        "      pgpKeyring = embed \"path/to/pgp-keyring\",\n"
```
Comment out IMHO since otherwise it won't compile

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "      # A keyring in GPG keyring format containing all public keys needed to verify PGP signatures in\n"
+        "      # this manifest (as of this writing, there is only one: `author.pgpSignature`).\n"
+        "      #\n"
+        "      # To generate a keyring containing just your public key, do:\n"
```
Keep this copy-pasta-type line IMHO

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "      #\n"
+        "      # Where `<key-id>` is a PGP key ID or email address associated with the key.\n"
+        "\n"
+        "      description = (defaultText = embed \"path/to/description.md\"),\n"
```
I think we should auto-generate a `.sandstorm/app-market-description.md` and then refer to it here, like what GitHub does for `README.md` on new repositories.

Else comment out.

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "      # in an app store. Note that the Markdown is not permitted to cotnain HTML nor image tags (but\n"
+        "      # you can include a list of screenshots separately).\n"
+        "\n"
+        "      shortDescription = (defaultText = \"one-to-three words\"),\n"
```
+1 keep this as-is

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "      screenshots = [\n"
+        "        # Screenshots to use for marketing purposes.  Examples below.\n"
+        "\n"
+        "        #(width = 746, height = 795, jpeg = embed \"path/to/screenshot-1.jpeg\"),\n"
```
I guess leave this commented out, yeah.

It is vaguely sad that people have to figure out the width & height of their images. ("If you took the screenshot on a Retina-type screen, divide the pixels by two" or something if you want.)

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "        #(width = 746, height = 795, jpeg = embed \"path/to/screenshot-1.jpeg\"),\n"
+        "        #(width = 640, height = 480, png = embed \"path/to/screenshot-2.png\"),\n"
+        "      ],\n"
+        "      changeLog = (defaultText = embed \"path/to/sandstorm-specific/changelog.md\"),\n"
```
Can we auto-generate an empty `app-market-changelog.md` like how GitHub does `README.md` in fresh git repositories? (Admittedly I'm suggesting we do it by default whereas for them it's an option.)

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -889,11 +889,99 @@ private:
+        "      ],\n"
+        "      changeLog = (defaultText = embed \"path/to/sandstorm-specific/changelog.md\"),\n"
+        "      # Documents the history of changes in Github-flavored markdown format (with the same restrictions\n"
+        "      # as govern `description`). We recommend formatting this with an H1 heading for each version\n"
```
Is it H1? I thought it was H3! Huh TIL.

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -918,12 +1006,66 @@ private:
           "  # `spk dev` will write a list of all the files your app uses to this file.\n"
           "  # You should review it later, before shipping your app.\n"
           "\n"
-          "  alwaysInclude = []\n"
```
+1 to trailing commas everywhere

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -918,12 +1006,66 @@ private:
+          "  #  # Uncomment this block and adjust the permissions and roles to make\n"
+          "  #  # sense for your app.\n"
+          "  #  # For more detailed information, see \"BridgeConfig\" in\n"
+          "  #  # https://github.com/sandstorm-io/sandstorm/blob/master/src/sandstorm/package.capnp\n"
```
Point to the authentication docs rather than `package.capnp`?

https://docs.sandstorm.io/en/latest/developing/auth/

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -918,12 +1006,66 @@ private:
+          "  #  # https://github.com/sandstorm-io/sandstorm/blob/master/src/sandstorm/package.capnp\n"
+          "  #  viewInfo = (\n"
+          "  #    # For details on the viewInfo field, consult \"ViewInfo\" in\n"
+          "  #    # https://github.com/sandstorm-io/sandstorm/blob/master/src/sandstorm/grain.capnp\n"
```
Same, the above doc hopefully answers these questions to avoid people needing to look at the capnp

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -918,12 +1006,66 @@ private:
+          "  #      # Roles are logical collections of permissions.  For instance, your app may have\n"
+          "  #      # a \"viewer\" role and an \"editor\" \n"
+          "  #      (\n"
+          "  #        title = (defaultText = \"admin\"),\n"
```
rename `admin` to `editor`

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -918,12 +1006,66 @@ private:
+          "  #        # Brief explanatory text to show in the sharing UI indicating\n"
+          "  #        # what a user assigned this role will be able to do with the grain.\n"
+          "  #\n"
+          "  #        description = (defaultText = \"admins may view all site data and change settings.\"),\n"
```
replace with "editors"

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -918,12 +1006,66 @@ private:
+          "  #        # Prose describing what this role means, suitable for a tool tip or similar help text.\n"
+          "  #      ),\n"
+          "  #    ],\n"
+          "  #  ),\n"
```
Provide an example of `viewer` and `editor` both in that case

paulproteus commented on src/sandstorm/spk.c++:
```diff
@@ -918,12 +1006,66 @@ private:
+          "  #      ),\n"
+          "  #    ],\n"
+          "  #  ),\n"
+          "  #  #apiPath = \"/api\",\n"
```
This is a pretty undocumented thing! I mean within this file. Add a comment maybe saying

```
# Apps can export an API to the world. The API is to be used by Javascript code and
# native apps, so it can't serve out regular HTML to browsers. If a request comes in to your app's API,
# sandstorm-http-bridge will prefix the request's path with this string.
```

something like that to indicate (a) what we mean by API ("oh this is for cap'nproto API exporting?") and (b) that the API not only shouldn't be used to expose your whole app, it _can't_ be effectively used that way ("oh I'll expose the whole app as an API so that I can bypass the Sandstorm sharing complexity")

kentonv commented:
Generally agreed with Asheesh's comments, except that I think linking to the source .capnp file (in addition to the docs) is a good thing: The source code is _precise_ documentation, whereas the docs are _readable_ documentation. IMO the latter cannot replace the former, and it frustrates me when people try to hide precise documentation in the name of readability.

ocdtrekkie commented:
FWIW, my docs article DOES actually link to the capnp file as well.

There seems to be a lot of mentions of the docs/capnp link in this. Could it be shortened a bit by just having the link to both once near the top and then below just saying 'refer to documentation' or similar?

zarvox commented:
Thanks for all the feedback, folks!

I aim to have each link point to the appropriate fragment in the documentation, so you can copy-paste the URL and be looking at the right thing immediately.  As the docs grow in detail and usefulness, the links to them need to grow increasingly precise to keep people from having to read a novel to find a piece of reference documentation.

I've attempted to reduce the repetition, but there's also some trickiness in that the .capnps relevant for developers span multiple files (and also can't link to fragments).

I'd love to be able to have links to specific capnp structs, and then we can point users at the high-level documentation, and the high-level docs can link to the precise docs in each paragraph, but I think that may require some additional tooling to make capnp documentation from the comments.

For now, I think this makes figuring out `metadata` and `bridgeConfig` way easier and makes @ocdtrekkie's excellent docs more discoverable, so I'm going to try not to let the perfect be the enemy of the good here.  We can write better guidance for how app developers should use `bridgeConfig` and connect the high-level docs to the precise docs as we continue forward. :)

paulproteus commented:
I'm +1 on this as-is, and I want to get feedback from people who use it and then have comments to drive further changes. I think @kentonv owns this code but it gets my +1 and I await his clicking "Merge"!
